Libraries/Opc.Ua.Security.Certificates/Extensions/X509SubjectAltNameExtension.cs

```diff
@@ -0,0 +1,446 @@
+using System.Text;
+
+namespace Opc.Ua.Security.Certificates
+{
+    /// <summary>
+    /// The subject alternate name extension.
+    /// </summary>
+    /// <remarks>
+    /// 
+    /// id-ce-subjectAltName OBJECT IDENTIFIER::=  { id-ce 17 }
+    /// 
+    /// SubjectAltName::= GeneralNames
+    /// 
+    ///    GeneralNames::= SEQUENCE SIZE(1..MAX) OF GeneralName
+    /// 
+    ///    GeneralName ::= CHOICE {
+    ///        otherName                       [0] OtherName,
+    ///        rfc822Name[1]                   IA5String,
+    ///        dNSName[2]                      IA5String,
+    ///        x400Address[3]                  ORAddress,
+    ///        directoryName[4]                Name,
+    ///        ediPartyName[5]                 EDIPartyName,
+    ///        uniformResourceIdentifier[6]    IA5String,
+    ///        iPAddress[7]                    OCTET STRING,
+    ///        registeredID[8]                 OBJECT IDENTIFIER
+    ///        }
+    /// 
+    ///    OtherName::= SEQUENCE {
+    ///        type-id                         OBJECT IDENTIFIER,
+    ///        value[0] EXPLICIT ANY DEFINED BY type - id
+    ///        }
+    /// 
+    ///    EDIPartyName::= SEQUENCE {
+    ///        nameAssigner[0]                 DirectoryString OPTIONAL,
+    ///        partyName[1]                    DirectoryString
+    ///        }
+    /// 
+    /// </remarks>
+    public class X509SubjectAltNameExtension : X509Extension
+    {
+        #region Constructors
+        /// <summary>
+        /// Creates an empty extension.
+        /// </summary>
+        protected X509SubjectAltNameExtension()
+        {
+        }
+
+        /// <summary>
+        /// Creates an extension from ASN.1 encoded data.
+        /// </summary>
+        public X509SubjectAltNameExtension(AsnEncodedData encodedExtension, bool critical)
+            : this(encodedExtension.Oid, encodedExtension.RawData, critical)
+        {
+        }
+
+        /// <summary>
+        /// Creates an extension from an Oid and ASN.1 encoded raw data.
+        /// </summary>
+        public X509SubjectAltNameExtension(string oid, byte[] rawData, bool critical)
+            : this(new Oid(oid, kFriendlyName), rawData, critical)
+        {
+        }
+
+        /// <summary>
+        /// Creates an extension from ASN.1 encoded data.
+        /// </summary>
+        public X509SubjectAltNameExtension(Oid oid, byte[] rawData, bool critical)
+        :
+            base(oid, rawData, critical)
+        {
+            m_decoded = false;
+        }
+
+        /// <summary>
+        /// Build the Subject Alternative name extension (for OPC UA application certs).
+        /// </summary>
+        /// <param name="applicationUri">The application Uri</param>
+        /// <param name="domainNames">The domain names. DNS Hostnames, IPv4 or IPv6 addresses</param>
+        public X509SubjectAltNameExtension(
+            string applicationUri,
+            IList<string> domainNames)
```

Comment:
Does it need to be IList? The other place I saw IList it could have been IEnumerable.

Reply:
Is this in general preferable to expose rather IEnumerable than IList for API? I can change it for sure here.